lib/include/jxl/encode.h

```diff
@@ -645,18 +645,99 @@ JXL_EXPORT JxlEncoderStatus JxlEncoderAddImageFrame(
+   *
+   * If the `size` argument points to 0 and the returned value is NULL, this
+   * will be interpreted as asking the output writing to stop. In such a case,
+   * the library will return an error. The client is expected to set the size of
+   * the returned buffer based on the suggested `size` when this function is
+   * called.
+   *
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param size points to a suggested buffer size when called; must be set to
+   * the size of the returned buffer once the function returns.
+   * @return a pointer to the acquired buffer or NULL to indicate an error or
+   * stop condition.
+   */
+  void* (*get_buffer)(void* opaque, size_t* size);
+
+  /**
+   * Required.
+   * Notifies the library that the current buffer's data has been written and
+   * can be released. This function should advance the current position of
+   * the buffer by `written_bytes` number of bytes.
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param written_bytes the number of bytes written to the buffer.
+   */
+  void (*release_buffer)(void* opaque, size_t written_bytes);
+
+  /**
+   * Optional, can be NULL
+   * Seeks to a specific position in the output. This function is optional and
+   * can be set to NULL if the output doesn't support seeking. Can only be done
+   * when there is no buffer.
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param opaque user supplied parameters to the callback
+   */
+  void (*seek)(void* opaque, uint64_t position);
+
+  /**
+   * Sets a watermark on the output data, at a specific position. This can be
+   * used to mark certain sections of the data for reference or for special
+   * processing.
+   *
+   * @param opaque user supplied parameters to the callback
+   * @param watermark_position the position, in bytes, where the watermark
+   * should be set.
+   */
+  void (*set_watermark)(void* opaque, uint64_t watermark_position);
+};
 
 /**
- * TODO(firsching): add documentation
+ * Sets the output processor for the encoder. This processor determines how the
+ * encoder will handle buffering, writing, seeking (if supported), and
+ * watermarking during the encoding process.
  *
+ * @param enc encoder object.
+ * @param output_processor the struct containing the callbacks for managing
+ * output.
+ * @return JXL_ENC_SUCCESS on success, JXL_ENC_ERROR on error.
  */
-JXL_EXPORT JxlEncoderStatus
-JxlEncoderSetOutputCallback(JxlEncoderOutputCallback callback);
+JXL_EXPORT JxlEncoderStatus JxlEncoderSetOutputProcessor(
+    JxlEncoder* enc, struct JxlEncoderOutputProcessor output_processor);
+
+/**
+ * Flushes any buffered input in the encoder, ensuring that all available input
+ * data has been processed and written to the output.
+ *
+ * @param enc encoder object.
+ * @return JXL_ENC_SUCCESS on success, JXL_ENC_ERROR on error.
+ */
+JXL_EXPORT JxlEncoderStatus JxlEncoderFlushInput(JxlEncoder* enc);
```

Comment:
We should mention how this function relates to ProcessOutput.

Also, we should mention that calling CloseInput should happen before the last call to FlushInput.

Reply:
mentioned it